src/XMakeBuildEngine/UnitTests/Evaluation/ImportFromMSBuildExtensionsPath_Tests.cs

```diff
@@ -143,25 +144,80 @@ public void ImportFromExtensionsPathWithWildCard()
-                                                    Console.WriteLine ("checking logcontains");
-                                                        l.AssertLogContains(String.Format(MockLogger.GetString("TargetDoesNotExist"), "FromExtn2"));
-                                                    });
+            CreateAndBuildProjectForImportFromExtensionsPath(mainProjectPath, "MSBuildExtensionsPath",
+                new[] {extnDir1, Path.Combine("tmp", "nonexistant"), extnDir2},
+                null,
+                (project, logger) =>
+                {
+                    Console.WriteLine(logger.FullLog);
+                    Console.WriteLine("checking FromExtn1");
+                    Assert.True(project.Build("FromExtn1"));
+                    Console.WriteLine("checking FromExtn2");
+                    Assert.True(project.Build("FromExtn2"));
+                    Console.WriteLine("checking logcontains");
+                    logger.AssertLogDoesntContain("MSB4057"); // Should not contain TargetDoesNotExist
+                });
+        }
+
+        [Fact]
+        public void ImportFromExtensionsPathWithWildCardAndSelfImport()
+        {
+            string mainTargetsFileContent = @"
+                <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                    <Target Name='Main'>
+                        <Message Text='Running Main'/>
+                    </Target>
+
+                    <Import Project='$(MSBuildExtensionsPath)\circularwildcardtest\*.proj'/>
+                </Project>";
+
+            string extnTargetsFileContent = @"
+                <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                    <Target Name='{0}'>
+                        <Message Text='Running {0}'/>
+                    </Target>
+                </Project>";
+
+            string extnTargetsFileContent2 = @"
+                <Project xmlns='http://schemas.microsoft.com/developer/msbuild/2003' >
+                    <Import Project='$(MSBuildExtensionsPath)\circularwildcardtest\*.proj'/>
+                    <Target Name='{0}'>
+                        <Message Text='Running {0}'/>
+                    </Target>
+                </Project>";
+
+            // Importing a wildcard will union all matching results from all fallback locations.
+            string extnDir1 = GetNewExtensionsPathAndCreateFile("extensions1", Path.Combine("circularwildcardtest", "extn.proj"),
+                string.Format(extnTargetsFileContent, "FromExtn1"));
+            string extnDir2 = GetNewExtensionsPathAndCreateFile("extensions2", Path.Combine("circularwildcardtest", "extn.proj"),
+                string.Format(extnTargetsFileContent, "FromExtn2"));
+            string extnDir3 = GetNewExtensionsPathAndCreateFile("extensions3", Path.Combine("circularwildcardtest", "extn3.proj"),
+                string.Format(extnTargetsFileContent2, "FromExtn3"));
```

Comment:
nit: using `extnTargetsFileContent`, same as for the other project files should be fine here too.